Calculate ahead-behind on worker thread

gerhardol opened the pull request:
Follow up to https://github.com/gitextensions/gitextensions/pull/11445#issuecomment-1861751454

## Proposed changes

Update after spinner is closed

Sidepanel calls are unchanged, FormBrowse is normally start first from what I see.

A second part could be to delay the start of the command until git-log delivers revisions.

## Test methodology <!-- How did you ensure quality? -->

Manual - inserted delay to the Git command

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

pmiossec commented:
@gerhardol I have pushed in this branch the other adjustments linked to ahead-behind I have. Please review.

If you don't want to see them in this branch, tell me I will do another PR.

gerhardol commented:
> If you don't want to see them in this branch, tell me I will do another PR.

We will have to agree regardless, so let us take the discussion here.

mstv commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -952,9 +954,18 @@ private void InternalInitialize()
-                    toolStripButtonPush.DisplayAheadBehindInformation(aheadBehindData, RevisionGrid.CurrentBranch.Value);
+                    if (AppSettings.ShowAheadBehindData)
+                    {
+                        this.InvokeAndForget(async () =>
+                        {
+                            bool isAheadBehindDisplayedForAllBranches = !MainSplitContainer.Panel1Collapsed && (AppSettings.RepoObjectsTreeShowRemotes || AppSettings.RepoObjectsTreeShowBranches);
+                            string branches = isAheadBehindDisplayedForAllBranches ? "" : RevisionGrid.CurrentBranch.Value;
+                            await TaskScheduler.Default;
+                            IDictionary<string, AheadBehindData> aheadBehindData = _aheadBehindDataProvider?.GetData(branches);
+                            await this.SwitchToMainThreadAsync();
+                            toolStripButtonPush.DisplayAheadBehindInformation(aheadBehindData, RevisionGrid.CurrentBranch.Value);
+                        });
```
```suggestion
                        bool isAheadBehindDisplayedForAllBranches = !MainSplitContainer.Panel1Collapsed && (AppSettings.RepoObjectsTreeShowRemotes || AppSettings.RepoObjectsTreeShowBranches);
                        string currentBranch = RevisionGrid.CurrentBranch.Value;
                        ThreadHelper.FileAndForget(async () =>
                        {
                            string aheadBehindBranch = isAheadBehindDisplayedForAllBranches ? "" : currentBranch;
                            IDictionary<string, AheadBehindData> aheadBehindData = _aheadBehindDataProvider?.GetData(aheadBehindBranch);
                            await this.SwitchToMainThreadAsync();
                            toolStripButtonPush.DisplayAheadBehindInformation(aheadBehindData, currentBranch);
                        });
```

mstv commented on GitCommands/Git/AheadBehindDataProvider.cs:
```diff
@@ -44,17 +46,21 @@ public void ResetCache()
+                // Callers setting branch name has the responsibility to ensure that not all are needed
+                if (string.IsNullOrWhiteSpace(branchName) && !string.IsNullOrWhiteSpace(_branchName))
+                {
+                    Trace.WriteLine($"Unexpectedly call for all branches after cache filled with specific branch {_branchName}");
```
```suggestion
                    Trace.WriteLine($"call for all branches after cache filled with specific branch {_branchName}");
```

mstv commented on GitCommands/Git/AheadBehindDataProvider.cs:
```diff
@@ -44,17 +46,21 @@ public void ResetCache()
-            return _lazyData.Value;
+                // Use Lazy<> to synchronize callers
+                _branchName = branchName;
+                Trace.WriteLine("_lazyData:" + _lazyData + "/bn:" + _branchName);
```
Use string interpolation and do not abbreviate?
```suggestion
                Trace.WriteLine($"{nameof(_lazyData)}: {_lazyData}, {nameof(_branchName)}: {_branchName}");
```

gerhardol replied:
Do you want to add this printout?

pmiossec replied:
No that's why I removed it when I was aware of this mistake.

mstv reviewed (commented):
nits

mstv reviewed (approved):
have not run

gerhardol commented on GitCommands/Git/AheadBehindDataProvider.cs:
```diff
@@ -44,17 +46,20 @@ public void ResetCache()
 
-            // Callers setting branch name has the responsibility to ensure that not all are needed
-            if (string.IsNullOrWhiteSpace(branchName) && !string.IsNullOrWhiteSpace(_branchName))
+            lock (_lock)
```
_lazyData could be init earler (but more complex), with the lock lazy could be removed.
Not worth changing...

mstv reviewed (approved):
👍

gerhardol commented:
Thee authors for one minor change!
Pushed an empty commit with suggested commit message at merge.

The size of the push button is briefly changed now.
We could set the text as ` ↑↓` temoporarily, to handle the common situation, or do as before and keep the old value.

Delaying ahead-behind could maybe be fixed, just a 200 ms delay or so. 
But maybe better to add a PostFirstRevisionRead event?

pmiossec commented:
> Three authors for one minor change! Pushed an empty commit with suggested commit message at merge. Edit: forgot co-author s

I squashed, improved commit message and added co-authors.

> The size of the push button is briefly changed now. We could set the text as ` ↑↓` temporarily, to handle the common situation, or do as before and keep the old value.

I don't see this size change on normal size repo and don't find it painful on a big repo. So OK for me at the moment.
We can reevaluate later. But if it's a pain, I think that we will have to retrieve ahead/behind only for the current branch to get the information immediately.

gerhardol commented:
> I squashed, improved commit message and added co-authors.

great
 
> > The size of the push button is briefly changed now. We could set the text as ` ↑↓` temporarily, to handle the common situation, or do as before and keep the old value.
> 
> I don't see this size change on normal size repo and don't find it painful on a big repo. So OK for me at the moment. We can reevaluate later. But if it's a pain, I think that we will have to retrieve ahead/behind only for the current branch to get the information immediately.

For me ahead-behind takes about 500ms.
I found that annoy, pushing a fixup. (I had hoped that keeping a same sized string would be good enough, but it was not). Some other  controls change size too.
No test added.
The alternative IMO is to keep the old behavior, keeping the old data until updating.
      
![push-empty](https://github.com/gitextensions/gitextensions/assets/6248932/166c82e5-2cd5-4a03-a6ef-1e7447358c16)

pmiossec commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -952,9 +953,18 @@ private void InternalInitialize()
-                    toolStripButtonPush.DisplayAheadBehindInformation(aheadBehindData, RevisionGrid.CurrentBranch.Value);
+                    if (AppSettings.ShowAheadBehindData)
+                    {
+                        bool isAheadBehindDisplayedForAllBranches = !MainSplitContainer.Panel1Collapsed && (AppSettings.RepoObjectsTreeShowRemotes || AppSettings.RepoObjectsTreeShowBranches);
+                        string currentBranch = RevisionGrid.CurrentBranch.Value;
+                        ThreadHelper.FileAndForget(async () =>
+                        {
+                            string aheadBehindBranch = isAheadBehindDisplayedForAllBranches ? "" : currentBranch;
+                            IDictionary<string, AheadBehindData> aheadBehindData = _aheadBehindDataProvider?.GetData(aheadBehindBranch);
+                            await this.SwitchToMainThreadAsync();
+                            toolStripButtonPush.DisplayAheadBehindInformation(aheadBehindData, currentBranch);
```
> For me ahead-behind takes about 500ms.
> I found that annoy, pushing a fixup.

If it is the case, please really consider doing the call here only for the current branch.
I know it will be an extra git call but it has the advantage to simplify the code (here and the new code you wish to add) and the call is pretty quick (60 to 90ms) so you have the benefit that the information displayed in the push button is always accurate (and it's a lot better when you open a big repository than having to wait more than 6s).
And maybe in this case, we could even remove the newly added `ThreadHelper.FileAndForget()` simplifying even more the code...

```suggestion
                        string currentBranch = RevisionGrid.CurrentBranch.Value;
                        ThreadHelper.FileAndForget(async () =>
                        {
                            IDictionary<string, AheadBehindData> aheadBehindData = _aheadBehindDataProvider?.GetData(currentBranch);
                            await this.SwitchToMainThreadAsync();
                            toolStripButtonPush.DisplayAheadBehindInformation(aheadBehindData, currentBranch);
```

pmiossec replied:
When testing this solution, I noticed there is a race condition:
When there is not a lot of local branches, the ahead-behind call for all branches arrives first and so execute quickly. The consequence is that the call specific for current branch is taking result from the cache and no 2nd git command is run.
When there are a lot of local branches (2k), the call for push button arrives first and so the lightest command is run first.

After testing with repositories with different branches count, it seems that the switch seems to happen when you have between 2k to 5k branches (where ahead-behind call takes more than 1.5s).
So, in the light of this information, I think we must:
* keep the `ThreadHelper.FileAndForget()` and your button size workaround contrary to what I just said earlier
* always try to get ahead-behind **for the current branch** here. If there are really a lot of branches, you will get the needed information quicker (because for me the ahead-behind in the push button is much more important than the one in the tree view). It's a kind of clue than the command will be long to run. In the other case, the command for the current branch won't be run and will take data from cache.

So I think you should accept my code suggestion and we are good to go.

mstv replied:
might be a usecase for `ExclusiveTaskRunner`

pmiossec replied:
I don't see a way how `ExclusiveTaskRunner` could improve the situation here. I think I understand what is the purpose but I don't see added value.

Except if we integrate all the retrieve thread logic into the provider with always launching thread to retrieve current branch before the one to retrieve all branches data. And we can start the threads earlier (just after retrieving data for revision grid) and accessing the data will wait completion of the thread(s).

A big refactoring but I don't think we will gain some time because the commands are launched pretty quickly in fact...

I think we are good enough now.

gerhardol replied:
The total time from the reset to the Git command is done is about 500 ms (so not just the Git command) - if the clear old data change is not reverted.
So current changes should be OK.

pmiossec commented:
FYI, I have pushed another little logical fix (even if I don't think it has any real effect)

gerhardol commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -955,12 +955,10 @@ private void InternalInitialize()
                         {
-                            string aheadBehindBranch = isAheadBehindDisplayedForAllBranches ? "" : currentBranch;
-                            IDictionary<string, AheadBehindData> aheadBehindData = _aheadBehindDataProvider?.GetData(aheadBehindBranch);
+                            IDictionary<string, AheadBehindData> aheadBehindData = _aheadBehindDataProvider?.GetData(currentBranch);
```
This is not something I prefer, it is OK to wait a little for this.
But I will not try block if the reason is well described in a comment.

Parallel git-for-each-ref will slow each other down.
Repeating that this info should not be requested until revisions are being shown (spinner is removed, Loading is shown in upper right corner of the grid.

pmiossec replied:
> This is not something I prefer, it is OK to wait a little for this.

As explained, this case will be beneficial when you have a lot of branches (maybe more than 2k) so in this exact case, you won't wait a little but maybe more than 1 or 2s (in my test repo it's more than 6s). So I thinks that better to not wait.

If you have less than this number of branches, in fact, whatever you pass as arguments (empty or current branch), the behavior will be the same and value will be taken from cache because the data retrieve is triggered first from left panel.

> Parallel git-for-each-ref will slow each other down.

And when you have a lot of branches and so there will 2 commands run, the first short lived one is already finished when the 2nd is launched.
So really no problem here....

As a summary, in normal case (not too much branch), the parameter is useless. And when we have a huge number of branches, it allows to get current branch info without slowing down retrieving data for all branches that happens later.

pmiossec replied:
@gerhardol have a look at the code comment to see if it is good enough...

gerhardol replied:
I can live with this.
But I will be annoyed next time I look at git commands and try to find what is not really needed and use a minute of my time to find the comment.

gerhardol replied:
I pushed a simple test case. @pmiossec merge if OK, you updated the commit message anyway

pmiossec replied:
> But I will be annoyed next time I look at git commands and try to find what is not really needed and use a minute of my time to find the comment.

It's what I try to explained, maybe not clearly enough, that in a normal size repository, the command **is not run at all** because the values are taken from the  filled dictionary because the retrieve from left panel that fetch **all** the data is done first. And the command is only run in huge repositories where it improve the UX a lot and without the 2 commands being run at the same time.

So you won't be annoyed 😉

gerhardol replied:
I will look at the log and try to optimize, will stop at the comment. So still some annoyed. But I would be more annoyed by running this alone.
Thanks for cooperation!
